### PR TITLE
GH-39707: [Java] Enable local build cache for Maven/Java build

### LINKED ISSUE #39707: [Java] Enable Java Build Caching
### Describe the enhancement requested

This enhancement was discussed at the [Dec 2023 Infra Roundtable](https://cwiki.apache.org/confluence/display/INFRA/Infra+Roundtable+2023-12-06%2C+17%3A00+UTC). Arrow was identified as a project that would be interested in build caching.

The proposed enhancement would connect the Java/Maven portion of the Arrow build to the Develocity instance at https://ge.apache.org. As a result of this, local build caching would become available to the Java/Maven portion of the Arrow build.

### Component(s)

Java

### PR DESCRIPTION
**Pull request by clayburn**
### Rationale for this change

This change has two main benefits:

#### Enabling local build caching

Enabling local build cache can speed up builds that occur on the same machine by skipping the execution of certain deterministic goals, such as Java compilation, when no change has occurred. 

In the future, [https://ge.apache.org](ge.apache.org) will support remote build caching as well so that results can be shared between builds (e.g. ephemeral CI build agents).

#### Enabling build scans

This change enables the publishing of build scans of the Apache Arrow project to the Develocity instance at [ge.apache.org](https://ge.apache.org/), hosted by the Apache Software Foundation and run in partnership between the ASF and Gradle. This Develocity instance has all features and extensions enabled and is freely available for use by the Apache Arrow project and all other Apache projects. Currently, Maven-built projects such as [Pulsar](https://ge.apache.org/scans?search.buildToolType=maven&search.rootProjectNames=*Pulsar*&search.timeZoneId=America%2FChicago), [IoTDB](https://ge.apache.org/scans?search.buildToolType=maven&search.rootProjectNames=*iotdb*&search.timeZoneId=America%2FChicago), [Ozone](https://ge.apache.org/scans?search.buildToolType=maven&search.rootProjectNames=*ozone*&search.timeZoneId=America%2FChicago), and others are using this instance.

On this Develocity instance, Apache Arrow will have access not only to all of the published build scans but other aggregate data features such as:

- Dashboards to view all historical build scans, along with performance trends over time
- Build failure analytics for enhanced investigation and diagnosis of build failures
- Test failure analytics to better understand trends and causes around slow, failing, and flaky tests

### What changes are included in this PR?

- Adds the [Develocity Maven Extension](https://docs.gradle.com/enterprise/maven-extension/) to enable local caching
- Publishes build scans to [ge.apache.org](https://ge.apache.org) from CI builds and authenticated local builds
- Adds the [Common Custom User Data Maven Extension](https://github.com/gradle/common-custom-user-data-maven-extension) to enhance build scans with more metadata

### Are these changes tested?

No (these changes are for the build and do not affect the code)

### Are there any user-facing changes?

No
* Closes: #39707

### CODE REVIEW

**Comment by danepitkin** [comment]
Thank you @clayburn! This is an exciting improvement. For local build execution, would we have to change anything in our developer docs[1] on how to build Arrow java in order to take advantage of this?

[1]https://arrow.apache.org/docs/developers/java/building.html#building-java-modules

**Comment by assignUser** [comment]
@clayburn I have approved the CI runs, I assume that only builds on main will be able to upload data due to the use of secrets? For similar things I have used a two step process in the past where the unprivileged workflow creates the data/artifacts and a second privileged workflow that doesn't run any user code would upload the data. Is this something gradle could support? Or maybe data for each PR build would be a bit much ^^

**Comment by clayburn** [comment]
> Thank you @clayburn! This is an exciting improvement. For local build execution, would we have to change anything in our developer docs[1] on how to build Arrow java in order to take advantage of this?
> 
> [1]https://arrow.apache.org/docs/developers/java/building.html#building-java-modules

@danepitkin - For caching, there is no action needed. The one stipulation that should be noted is that the local cache is machine local, so with some of your containerized builds, it would be local to the container itself without any special handling. The local cache by default is located at `$HOME/.m2/.gradle-enterprise/build-cache`.

For build scans, you won't see any changes unless you explicitly authenticate to https://ge.apache.org. Instructions to do so are [here](https://docs.gradle.com/enterprise/maven-extension/#authenticating_with_gradle_enterprise). All ASF committers can authenticate to ge.apache.org with LDAP credentials. Contributors cannot authenticate, so they won't see anything regarding build scans. The access key is either stored in `~/.m2` or pulled from an env var, so again, the containerized builds may not authenticate easily without intention from the user.

> @clayburn I have approved the CI runs, I assume that only builds on main will be able to upload data due to the use of secrets? For similar things I have used a two step process in the past where the unprivileged workflow creates the data/artifacts and a second privileged workflow that doesn't run any user code would upload the data. Is this something gradle could support? Or maybe data for each PR build would be a bit much ^^

Build scans will only be produced for builds that run from the `apache/arrow` repo. In other words, not from forks. Forked builds will still build just fine, but will not produce a scan. We are working on methods to handle this in the way you described, performing the upload of the scan in a "trusted" environment, but this is not yet ready.

The caching functionality will work without the secret, although the benefits may not be apparent in your CI builds since they are presumably running on ephemeral agents and the cache is machine local. This is where the remote cache will help more, which we hope to enable in the future.

**Comment by pitrou** [comment]
> The local cache by default is located at `$HOME/.m2/.gradle-enterprise/build-cache`.

Can it be directed to another directory using an environment variable? In C++ we direct ccache data to a [custom directory](https://github.com/apache/arrow/blob/c33ffb033a49c17619f2547d873569980ee91b9c/docker-compose.yml#L61-L66) which is then persisted as [Docker volumes](https://github.com/apache/arrow/blob/c33ffb033a49c17619f2547d873569980ee91b9c/docker-compose.yml#L178-L200). This is then used together with [GHA caching](https://github.com/apache/arrow/blob/c33ffb033a49c17619f2547d873569980ee91b9c/.github/workflows/cpp.yml#L103-L108) to shorten CI times.

**Comment by clayburn** [comment]
> > The local cache by default is located at `$HOME/.m2/.gradle-enterprise/build-cache`.
> 
> Can it be directed to another directory using an environment variable? In C++ we direct ccache data to a [custom directory](https://github.com/apache/arrow/blob/c33ffb033a49c17619f2547d873569980ee91b9c/docker-compose.yml#L61-L66) which is then persisted as [Docker volumes](https://github.com/apache/arrow/blob/c33ffb033a49c17619f2547d873569980ee91b9c/docker-compose.yml#L178-L200). This is then used together with [GHA caching](https://github.com/apache/arrow/blob/c33ffb033a49c17619f2547d873569980ee91b9c/.github/workflows/cpp.yml#L103-L108) to shorten CI times.

I think this is plausible. I'll give it a shot and update back here.

**Comment by clayburn** [comment]
@pitrou - Looking through the workflow, you are already [caching the right directory](https://github.com/apache/arrow/blob/main/.github/workflows/java.yml#L71). It's all of `$HOME/.m2`, but it will include the default build cache directory. The cache key there being constructed by all files under `java/**` may be overly specific for the build cache, but the `restore-keys` configuration should handle it. We can see as you publish build scans and generate that data.

I've pushed two changes to further optimize this:
- One that enables the local cache. It also adds the clean lifecycle to the CI invocations of Maven, as the build cache is in read-only mode when clean isn't used.
- One that defines some normalizations so that, for example, a change to a properties file that exists in a runtime classpath won't cause a cache miss.

File: .mvn/gradle-enterprise.xml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        <obfuscation>
+            <ipAddresses>#{{'0.0.0.0'}}</ipAddresses>
+        </obfuscation>
+    </buildScan>
+    <buildCache>
+        <remote>
+            <enabled>false</enabled>
+        </remote>
+    </buildCache>
+</gradleEnterprise>
```
Comment (jduo) [inline .mvn/gradle-enterprise.xml:45]: Please add newlines to EOF

File: ci/scripts/java_build.sh [hunk 2]
```diff
@@ -89,12 +89,12 @@ if [ "${ARROW_JAVA_JNI}" = "ON" ]; then
...
 fi
 
-${mvn} install
+${mvn} clean install
 
 if [ "${BUILD_DOCS_JAVA}" == "ON" ]; then
   # HTTP pooling is turned of to avoid download issues https://issues.apache.org/jira/browse/ARROW-11633
   mkdir -p ${build_dir}/docs/java/reference
-  ${mvn} -Dcheckstyle.skip=true -Dhttp.keepAlive=false -Dmaven.wagon.http.pool=false install site
+  ${mvn} -Dcheckstyle.skip=true -Dhttp.keepAlive=false -Dmaven.wagon.http.pool=false clean install site
```
Comment (lidavidm) [inline ci/scripts/java_build.sh:97]: I suppose build caching means that the 'clean' here isn't actually a clean build?

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

**Review by danepitkin — APPROVED** [review]
Thanks for the contribution!

**Comment by clayburn** [comment]
@lidavidm Done ✅

**Comment by clayburn** [comment]
> > For caching, there is no action needed. The one stipulation that should be noted is that the local cache is machine local
> 
> @clayburn If you could help me with doubts regarding local development, I would appreciate it:
> 
> 1. I am running `$ mvn clean install` and seeing that `/.m2/.gradle-enterprise` is populating with information. If I run `$ mvn clean install` again and haven't made any changes to my Java module projects, should it use the last cache?
> 2. Is there a list of tasks that will be fetched from the cache before building it again (e.g. compile,...)?
> 
> I would like to identify/map what the new advantages will be for developers on their loca machines.
> 
> Thank you in advance for your support.

@davisusanibar - So sorry for losing track of this:

1. Yes, such a build will use the cache. It also isn't necessarily even the "last" cache (i.e. the last build). It could be the case that you use cache entries from older builds as well. This can help, for example, when switching back and forth between feature branches.

2. We document the default cacheable goals [here](https://docs.gradle.com/enterprise/maven-extension/#cacheable_plugins_and_goals). It is possible to add goals if needed. We can explore this in the future to see if any in your project make sense to add.

**Comment by danepitkin** [comment]
@github-actions crossbow submit -g java

File: .github/workflows/java.yml [hunk 3]
```diff
@@ -84,7 +84,8 @@ jobs:
         env:
           ARCHERY_DOCKER_USER: ${{ secrets.DOCKERHUB_USER }}
           ARCHERY_DOCKER_PASSWORD: ${{ secrets.DOCKERHUB_TOKEN }}
-        run: archery docker run ${{ matrix.image }}
+          GRADLE_ENTERPRISE_ACCESS_KEY: ${{ secrets.GE_ACCESS_TOKEN }}
+        run: archery docker run ${{ matrix.image }} -e "GRADLE_ENTERPRISE_ACCESS_KEY=$GRADLE_ENTERPRISE_ACCESS_KEY" -e CI=true
```
Comment (kou) [inline .github/workflows/java.yml:88]: Could you use `archery docker run -e ... ${{ matrix.image }}` style like we did in other places?

```suggestion
        run: |
          archery docker run \
            -e CI=true \
            -e "GRADLE_ENTERPRISE_ACCESS_KEY=$GRADLE_ENTERPRISE_ACCESS_KEY" \
            ${{ matrix.image }}
```

File: .github/workflows/java_jni.yml [hunk 4]
```diff
@@ -118,7 +118,8 @@ jobs:
         env:
           ARCHERY_DOCKER_USER: ${{ secrets.DOCKERHUB_USER }}
           ARCHERY_DOCKER_PASSWORD: ${{ secrets.DOCKERHUB_TOKEN }}
-        run: archery docker run conda-python-java-integration
+          GRADLE_ENTERPRISE_ACCESS_KEY: ${{ secrets.GE_ACCESS_TOKEN }}
+        run: archery docker run conda-python-java-integration -e "GRADLE_ENTERPRISE_ACCESS_KEY=$GRADLE_ENTERPRISE_ACCESS_KEY" -e CI=true
```
Comment (kou) [inline .github/workflows/java_jni.yml:122]: ditto.

**Comment by lidavidm** [comment]
@danepitkin @vibhatha A potential CI improvement might be to have a post-run step that captures and uploads dump files like these so we can get to the bottom of this instability.

```
Error:  Caused by: org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
Error:  Command was cmd.exe /X /C "C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\11.0.22-7\x64\bin\java --add-opens=java.base/java.nio=org.apache.arrow.memory.core,ALL-UNNAMED -jar C:\Users\runneradmin\AppData\Local\Temp\surefire832983470661055534\surefirebooter-20240322025403947_28.jar C:\Users\runneradmin\AppData\Local\Temp\surefire832983470661055534 2024-03-22T02-53-37_914-jvmRun2 surefire-20240322025403947_24tmp surefire_6-20240322025403947_26tmp"
```

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Verification failure is unrelated. However, another improvement might be to run `mvn -B` in our verification script; I don't think we need to see the entire download progress in logs and it makes scanning the log much harder...

**Comment by danepitkin** [comment]
Hmm, I mightve found the culprit.. DEBUG messages are enabled in the build now causing a huge amount of logging.

**Comment by lidavidm** [comment]
Do we want to revert, or do you think disabling the log is good enough?

**Comment by danepitkin** [comment]
I think disabling DEBUG logs should be good enough. I'll try it out.

**Pull request merged**